### PR TITLE
Fixes for julia v1.10

### PR DESCRIPTION
**Pull request by IanButterworth**
Proposed fix for #108 

Includes the fix from #112 to try and get CI green

@vtjnash

### CODE REVIEW

File: src/reflection/utils.jl [hunk 1]
```diff
@@ -128,7 +128,11 @@ function splicearg!(ir::IR)
   return arg
 end
 
-@static if VERSION < v"1.8.0-DEV.267"
+@static if VERSION >= v"1.10.0-DEV.870"
+  function replace_code_newstyle!(ci, ir, _)
+    return Core.Compiler.replace_code_newstyle!(ci, ir)
```
Comment (vtjnash) [inline src/reflection/utils.jl:133]: you also might need to allocate slottypes, if you didn't already do that
```
    ci.slottypes isa Array || (ci.slottypes = [])
```

File: src/reflection/reflection.jl [hunk 2]
```diff
@@ -54,16 +54,17 @@ function meta(T; types = T, world = worldcounter())
...
     mi = Core.Compiler.code_for_method(method, types, sps, world, false)
-    ci = hasgenerator(mi) ? Core.Compiler.get_staged(mi) : Base.uncompressed_ast(mi)
+    ci = hasgenerator(mi) ? Core.Compiler.get_staged(mi, world) : Base.uncompressed_ast(mi)
   end
   Base.Meta.partially_inline!(ci.code, [], method.sig, Any[sps...], 0, 0, :propagate)
   Meta(method, mi, ci, method.nargs, sps)
 end
 
 function invoke_tweaks!(ci::CodeInfo)
+  ci.slottypes = [:todo, ci.slottypess[1], :todo, ci.slottypes[2:end]...]
```
Comment (IanButterworth) [inline src/reflection/reflection.jl:67]: I didn't know what to put here where there is `:todo`

Comment (IanButterworth) [reply src/reflection/reflection.jl]: I've gone with
```
  ci.slottypes = [typeof(invoke), ci.slottypess[1], Type, ci.slottypes[2:end]...]
```

**Comment by IanButterworth** [comment]
I think I've done all the obvious stuff here. Seems like it needs someone with more IRTools knowledge

**Comment by IanButterworth** [comment]
@aviatesk @vtjnash any idea what's going on here? I'm not familiar with IRTools at all

**Comment by vtjnash** [comment]
Oh, it is using APIs that Base doesn't have anymore and need to be converted to the new API

**Comment by maleadt** [comment]
Here's with those APIs fixed: https://github.com/FluxML/IRTools.jl/compare/master...maleadt:IRTools.jl:replace_code_newstyle_fix
@IanButterworth Can you merge that here?

Only a couple of test failures remain.

**Comment by IanButterworth** [comment]
```
Compiler: Test Failed at /home/runner/work/IRTools.jl/IRTools.jl/test/compiler.jl:93
  Expression: mullify(prod, [5, 10]) == 15
   Evaluated: 50 == 15
```
Is because https://github.com/IanButterworth/IRTools.jl/blob/668530fb38ca20678e0934efb3b74d93d8711925/test/compiler.jl#L80 is returning as nothing

**Comment by IanButterworth** [comment]
Appears to be a world age issue
```
julia> @dynamo function mullify(a...)
         @show ir = IR(a...)
         ir == nothing && return
         ir = MacroTools.prewalk(ir) do x
           x isa GlobalRef && x.name == :(*) && return GlobalRef(Base, :+)
           return x
         end
         for (x, st) in ir
           isexpr(st.expr, :call) || continue
           ir[x] = xcall(Main, :mullify, st.expr.args...)
         end
         return ir
       end

julia> mullify(prod, [5, 10])
Tuple{Ts...} = Tuple{typeof(prod), Vector{Int64}}
m = IRTools.meta(Tuple{Ts...}) = nothing
ir = IR(a...) = nothing

julia> IRTools.meta(Tuple{typeof(prod), Vector{Int64}})
Metadata for prod(a::AbstractArray; dims, kw...) @ Base reducedim.jl:994
```

**Comment by maleadt** [comment]
> Might this be related to Jameson's comment here [f401a95#r111498985](https://github.com/FluxML/IRTools.jl/commit/f401a9596921a1f30c5f4a15e77cad4d2e8e1ae1#r111498985) @maleadt ?

No, the problem is that `@dynamo` is fundamentally broken. Or at least the cases where the dynamo looks up Julia IR, which now needs to be done with world ages in mind (i.e. passing the age to reflection functions and setting edges on the returned CodeInfo), like the `mullify` one above. Sadly, this also seems like the way most packages use `@dynamo`.

**Comment by maleadt** [comment]
Updated my fork; it passes tests now.

**Comment by IanButterworth** [comment]
Great! Although Zygote still fails to precompile on https://github.com/FluxML/Zygote.jl/blob/2287a86d1692c0dfc568739a80ca909199f6adfe/src/precompile.jl#LL17C20-L17C20

But perhaps that shouldn't stop this being merged and released?
```
ERROR: LoadError: MethodError: no method matching iterate(::Nothing)

Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen})
   @ Base range.jl:887
  iterate(::Union{LinRange, StepRangeLen}, ::Integer)
   @ Base range.jl:887
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}}
   @ Base dict.jl:728
  ...

Stacktrace:
  [1] indexed_iterate(I::Nothing, i::Int64)
    @ Base ./tuple.jl:95
  [2] chain_rrule(::Zygote.ZygoteRuleConfig{Zygote.Context{false}}, ::typeof(Zygote.pow), ::Int64, ::Int64)
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/chainrules.jl:223 [inlined]
  [3] macro expansion
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/interface2.jl:0 [inlined]
  [4] _pullback(::Zygote.Context{false}, ::typeof(Zygote.pow), ::Int64, ::Int64)
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/interface2.jl:9
  [5] pullback(::Function, ::Zygote.Context{false}, ::Int64, ::Vararg{Int64})
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/interface.jl:44
  [6] pullback(::Function, ::Int64, ::Int64)
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/interface.jl:42
  [7] gradient(::Function, ::Int64, ::Vararg{Int64})
    @ Zygote ~/.julia/packages/Zygote/SuKWp/src/compiler/interface.jl:96
  [8] top-level scope
    @ ~/.julia/packages/Zygote/SuKWp/src/precompile.jl:17
```

**Comment by maleadt** [comment]
I pushed another fix to my branch.

Some initial fixes for Zygote.jl are in https://github.com/maleadt/Zygote.jl/tree/dev, depending on this PR. This makes Zygote precompile and makes a smoke test `gradient` succeed, but there's more instances where the generated world age isn't propagated to reflection methods (resulting in `nothing` results, or in `code reflection cannot be used from generated functions` errors). Those should be relatively easy to fix though, with the core fixes in place.

**Comment by IanButterworth** [comment]
It would be good to get this released

**Comment by maleadt** [comment]
@IanButterworth Can you merge the latest version of my branch?

**Comment by IanButterworth** [comment]
I believe this branch is up to date with yours?

**Comment by maleadt** [comment]
Ah yes, the GitHub UI is confusing here.

Can we get this (squash-)merged and tagged then? @CarloLucibello @ToucheSir

**Pull request merged**